compiler/resolution/lowerForalls.cpp

```diff
@@ -873,6 +893,89 @@ static void expandTopLevel(ExpandVisitor* outerVis,
 }
 
 
+/////////// reorder shadow variables ///////////
+
+// "Proper" shadow variables i.e. not TPVs should go first,
+// except those for reduce intents.
+static bool shouldGoFirst(ShadowVarSymbol* sv) {
+  switch (sv->intent)
+  {
+  case TFI_DEFAULT:
+  case TFI_CONST:
+  case TFI_IN_OUTERVAR:
+  case TFI_IN:
+  case TFI_CONST_IN:
+  case TFI_REF:
+  case TFI_CONST_REF:
+    return true;
+
+  case TFI_REDUCE:
+  case TFI_REDUCE_OP:
+  case TFI_TASK_PRIVATE:
+    return false;
+  }
+
+  return false; // dummy
+}
+
+//
+// We want all proper shadow variables to be ahead of task-private variables.
+// Explanation: (a) The latter may reference the former - in TPVs initBlocks.
```

Comment:
It might help to have a code example showing how one can refer to the other and another example showing how the other direction is not possible.